File: 代码/leetcodeP/grady/leet2193.py
"""
@author:fuzekun
@file:leet2193.py
@time:2022/08/10
@description:
得到回文串的最少操作次数
"""
import collections
from collections import Counter
from collections import defaultdict
from dataStructure import treeArray


class Solution:
    def minMovesToMakePalindrome(self, s: str) -> int:
        s = list(s)
        n = len(s)
        i, j, ans = 0, n - 1, 0
        while i < j:                    # 由于range生成的是不可变序列所以使用while循环
            k = j
            ch = s[i]
            while s[k] != ch :
                k -= 1
            if k != i:                  # 如果出现了偶数次
                ans += j - k
                while k < j:
                    s[k], s[k + 1] = s[k + 1], s[k]
                    k += 1
                j -= 1
            else :                      # 如果出现了奇数次
                ans += n // 2 - i
            i += 1
        return ans

    def minMove2(self, s: str) -> int :
        n = len(s)
        # 进行一个计数, 并且得到是奇数的情况
        freq = Counter(c for c in s)
        jich = [ch for ch in freq if freq[ch] % 2 == 1]
        # 获取左边编码，右边下标，以及组间间距
        lcnt, rcnt = 0, 0
        left, right = defaultdict(list), defaultdict(list)
        ans = 0
        for i, ch in enumerate(s) :
            if left[ch].__len__() < freq[ch] // 2:
                lcnt += 1
                left[ch].append(lcnt)
                ans += i - lcnt + 1
            else :
                rcnt += 1
                right[ch].append(rcnt)

        # 处理奇数的情况
        if len(jich) != 0:
            lcnt += 1
            left[jich[0]].append(lcnt)
        # 生成右边的排列
        perm = [0] * ((n - 1) // 2 + 1)
        for ch, rlst in right.items() :
            llst = left[ch]
            m = len(rlst)
            for i in range(m) :
                perm[rlst[m - 1 - i] - 1] = llst[i]
        print(perm)
        perm = reversed(perm)
        # 求解逆序对
        maxv = 1005
        ta = treeArray.TreeArray(maxv)
        for x in perm:
           ans += ta.sum(maxv) - ta.sum(x)
           ta.add(x, 1)

        return ans


if __name__ == '__main__':
    s = "aaabbbbcc"
    su = Solution()
    ans = su.minMove2(s)
    print("ans = " + str(ans))

File: codes/acwing/src/main/java/grady/leet2193.java
package grady;

import dataStructure.TreeArray;
import org.junit.Test;

import java.util.*;

/**
 * @author: Zekun Fu
 * @date: 2022/8/9 9:27
 * @Description: 得到会问串的最少操作次数
 *
 * 1, 出现奇数次的最多有一种。
 * 2. 出现偶数次的回文字母，相对顺序一定保持不变。
 *
 *
 * 有一个问题，会不会，一旦一个字母交换完成，另一个字母就受到影响。
 * 会不会和字母的交换顺序有什么关系呢？
 *
 * 所以就有了下面的构造方法：
 * 一、对于偶数次的字母
 * 1. 对于任意一个字母， 采用双指针。
 * 2. 遍历左半部分i，计算右半部分对应的位置p = n - i - 1 。
 *3. 右指针从右向左进行，计算需要交换的次数. ans += abs(p - k)
 *
 * 二、对于奇数的字母
 * 直接删掉，就变成了子问题。最后把它加进去需要n / 2 - i次交换
 *
 * aaabb aabba abba ababa
 *
 *
 *
 * 问题
 *1. 为什么要贪心
 *2. 代码中为什么右边的序列里面要和左边的逆序匹配
 *
 *
 *
 *
 */
public class leet2193 {
    public int minMovesToMakePalindrome(String ss) {
        char[] s = ss.toCharArray();
        int n = s.length, ans = 0;
        for (int i = 0, j = n - 1; i < j; i++) {
            boolean flag = false;
            for (int k = j; i != k; k--) {
                if (s[i] == s[k]) {
                    for (; k < j; k++) {
                        char tmp = s[k];
                        s[k] = s[k + 1];
                        s[k + 1] = tmp;
                        ans++;
                    }
                    j--;
                    flag = true;
                    break;
                }
            }
            if (!flag) ans += n / 2 - i;
        }
        return ans;
    }


    public int m2(String ss) {
        char[] s = ss.toCharArray();
        Map<Character, Integer> freq = new HashMap<>();
        for (char ch :s) {
            if (freq.containsKey(ch)) freq.put(ch, freq.get(ch) + 1);
            else freq.put(ch, 1);
        }

        int ans = 0;
        // 存储每一个字符对应的编码, 从1开始编码
        HashMap<Character, List<Integer>> left = new HashMap<>();
        Map<Character, List<Integer>>right = new HashMap<>();
        int lcnt = 0, rcnt = 0;
        int n = s.length;
        for (int i = 0; i < n; i++) {
            char ch = s[i];
            left.putIfAbsent(ch, new ArrayList<>());
            right.putIfAbsent(ch, new ArrayList<>());
            List<Integer>llist = left.get(ch);
            List<Integer> rlist = right.get(ch);
            if (llist.size() < freq.get(ch) / 2) {
                ++lcnt;
                llist.add(lcnt);
                ans += (i - lcnt + 1);  // 计算组间间距
            } else {
                ++rcnt;
                rlist.add(rcnt);
            }
        }

        // 处理奇数的情况
        if (n % 2 == 1) {
            for (Map.Entry<Character, Integer> entry : freq.entrySet()) {
                char c = entry.getKey();
                int occ = entry.getValue();
                if (occ % 2 == 1) {
                    ++lcnt;
                    left.get(c).add(lcnt);
                    break;
                }
            }
        }
        for (Map.Entry<Character, List<Integer>>entry :left.entrySet()) {
            System.out.println("ch = " + entry.getKey());
            System.out.println(entry.getValue());
        }
        for (Map.Entry<Character, List<Integer>>entry :right.entrySet()) {
            System.out.println("ch = " + entry.getKey());
            System.out.println(entry.getValue());
        }

        // 生成右边的排列
        int[] perm = new int[(n + 1) / 2];
        for (Map.Entry<Character, List<Integer>> entry : right.entrySet()) {
            List<Integer>rlist = entry.getValue();
            List<Integer>llist = left.get(entry.getKey());
            int m = rlist.size();
            for (int i = 0; i < m; i++) {
                perm[rlist.get(m - i - 1) - 1] = llist.get(i);
            }
        }
        for (int x :perm) {
            System.out.print( x + " ");
        }
        System.out.println();
        for (int i = 0, j = perm.length - 1; i < j; i++, j--) {
            int tmp = perm[i];
            perm[i] = perm[j];
            perm[j] = tmp;
        }

        //  使用树状数组计算逆序对
        int maxv = 1005;
        TreeArray ta = new TreeArray(maxv); // 最多1000
        for (int x : perm) {

            ans += ta.sum(maxv) - ta.sum(x); // (x, maxv]
            ta.add(x, 1);
        }
        System.out.println();
        return ans;
    }

    @Test
    public void testM2() {
        String s = "aaabbbbcc";
        int ans = m2(s);
        System.out.println("ans = " + ans);
    }
}
